signac/core/collection_api.py

```diff
@@ -0,0 +1,155 @@
+# Copyright (c) 2020 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+from contextlib import contextmanager
+from abc import abstractmethod
+from abc import ABCMeta
+from collections import defaultdict
+from collections.abc import Collection
+
+try:
+    import numpy
+    NUMPY = True
+except ImportError:
+    NUMPY = False
+
+
+class SyncedCollectionABCMeta(ABCMeta):
+    """ Metaclass for the definition of SyncedCollection.
+
+    This metaclass automatically registers synced data structures' definition,
+    this is used when recursively converting synced data structures to determine
+    what to convert their children into.
```

Comment:
It is still not clear to me what the purpose of the meta class is or how I as a developer would need to use it. Specifically I don't understand what a "synced data structure definition" is.

Meta classes are a more advanced and somewhat abstract language feature and should therefore be used with caution, because it reduces code clarity and limits who can maintain this code. So if we are using meta classes, then we should have strong justification for it and on top need to document extremely well what the purpose is and how they work.

I have admittedly not followed the full discussion leading up to the class structure that is implemented here and forgot the details of the proposal, however I am otherwise very familiar with the code base and the general purpose of these classes and I think it is a bit of a red flag that I can't fully follow what is happening here. So we definitely need to improve the code documentation here, but we should also consider whether we can achieve the same thing maybe in a more obvious way without meta classes.

Reply:
I'm very open to finding a simpler solution if you can think of one. The problem we're attempting to solve is the following. The open "hard-to-solve" issues regarding synced dicts in the current codebase generally have to do with nesting synced dicts/lists and finding that things break. The metaclass approach is designed to solve this by defining the conversion from base collections to the corresponding synced collections as a recursive conversion that searches for the appropriate corresponding type in the registry of available synced collections. This ensures that a `dict[dict[list]]` doesn't lead to a `SyncedDict[SyncedDict[list]]` instead of `SyncedDict[SyncedDict[SyncedList]]`. The main reason the metaclass is necessary (and I'm very open to alternate solutions) is to enable the flexibility for users to define their own collections after importing the class and ensuring that the appropriate types of objects are created by the recursion. An auxiliary benefit is that it solves any circular import issues that might otherwise arise in the package where the SyncedCollection class would need to know which of its children should be instantiated in any particular case. The latter problem could probably be solved in other ways, but I'm not sure if there's a simpler or more elegant solution than a metaclass for the former.

Reply:
Unless I'm missing some kind of obvious flaw with this, one very simple alternative approach would be to simply register those classes with the base class with a class attribute and method:
```python
JsonCollection.register(JsonDict)
JsonCollection.register(JsonList)
# ...
```
Use `*args` if you want this to be more concise:
```python
JsonCollection.register(JsonDict, JsonList, ...)
```

Reply:
Okay, I will try that.

Reply:
I have done the changes.

Reply:
@csadorf I don't see any significant flaws in that approach, other than the fact that it just doesn't handle this automatically. I'm fine with choosing the more explicit approach here for simplicity.

Reply:
I'm reviewing this PR now. I don't have strong opinions about metaclass vs. explicit registration. I think this conversation can be considered resolved.